### PR TITLE
Nextcloud native on NGINX

### PR DESCRIPTION
**Pull request by holta**
This PR:
- Fixes Nextcloud icons thanks to @m-anish who trialed that a month ago with PR #2119.
- Installs Nextcloud's "official" php modules per https://docs.nextcloud.com/server/18/admin_manual/installation/source_installation.html#prerequisites-for-manual-installation
- Mitigates most all of the yellow/orange-colored Nextcloud security/configuration warnings (http://box/nextcloud -> A [top-right] -> Settings -> Overview [left side]) thanks to improvements in nextcloud/templates/nextcloud-nginx.conf.j2
- Implements nginx_high_php_limits in /etc/php/7.x/fpm/php.ini, in the place of apache_high_php_limits
- Disables var nextcloud_allow_public_ips as this might prove unneeded with NGINX
- Installs php-fpm as Debian (and others?) require, for now in Stage 3's roles/www_back_end [since moved to Stage 4's roles/www_front_end]
- Takes a 1st stab at updating [roles/nextcloud/README.md](https://github.com/holta/iiab/blob/nextcloud-native-on-nginx/roles/nextcloud/README.md) documentation (further updates expected in Q1 2020 as this all comes together!)

Tested on Debian 10.3 & Ubuntu 19.10

### CODE REVIEW

**Comment by holta** [comment]
TO DO:
- Try to move /library/www/html/nextcloud/data to /library/nextcloud/data [or perhaps better yet outside of /library -- since personal/private data very likely doesn't belong in a "public library"]
- Test with a fresh install on RPi

**Comment by holta** [comment]
Indeed testing on Debian & Ubuntu was not quite enough, as suspected.

But hopefully this Raspbian issue can be quickly solved Sunday:

> TASK [nextcloud : Install ffmpeg + libxml2 + 21 php packages] ******************
fatal: [127.0.0.1]: FAILED! => {"cache_update_time": 1581806445, "cache_updated": false, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"      install 'ffmpeg' 'php-bz2' 'php-ctype' 'php-curl' 'php-dom' 'php-exif' 'php-fileinfo' 'php-gd' 'php-gmp' 'php-iconv' 'php-intl' 'php-json' 'php-mbstring' 'php-mysql' 'php-posix' 'php-simplexml' 'php-smbclient' 'php-xmlreader' 'php-xmlwriter' 'php-zip'' failed: E: Package 'php-ctype' has no installation candidate\nE: Package 'php-dom' has no installation candidate\nE: Package 'php-exif' has no installation candidate\nE: Package 'php-fileinfo' has no installation candidate\nE: Package 'php-iconv' has no installation candidate\nE: Package 'php-posix' has no installation candidate\nE: Package 'php-simplexml' has no installation candidate\nE: Package 'php-xmlreader' has no installation candidate\nE: Package 'php-xmlwriter' has no installation candidate\n", "rc": 100, "stderr": "E: Package 'php-ctype' has no installation candidate\nE: Package 'php-dom' has no installation candidate\nE: Package 'php-exif' has no installation candidate\nE: Package 'php-fileinfo' has no installation candidate\nE: Package 'php-iconv' has no installation candidate\nE: Package 'php-posix' has no installation candidate\nE: Package 'php-simplexml' has no installation candidate\nE: Package 'php-xmlreader' has no installation candidate\nE: Package 'php-xmlwriter' has no installation candidate\n", "stderr_lines": ["E: Package 'php-ctype' has no installation candidate", "E: Package 'php-dom' has no installation candidate", "E: Package 'php-exif' has no installation candidate", "E: Package 'php-fileinfo' has no installation candidate", "E: Package 'php-iconv' has no installation candidate", "E: Package 'php-posix' has no installation candidate", "E: Package 'php-simplexml' has no installation candidate", "E: Package 'php-xmlreader' has no installation candidate", "E: Package 'php-xmlwriter' has no installation candidate"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nPackage php-ctype is a virtual package provided by:\n  php7.3-common 7.3.11-1~deb10u1\n  php7.2-common 7.2.9-1+b2\n  php7.1-common 7.1.20-1+b2\n\nPackage php-exif is a virtual package provided by:\n  php7.3-common 7.3.11-1~deb10u1\n  php7.2-common 7.2.9-1+b2\n  php7.1-common 7.1.20-1+b2\n\nPackage php-fileinfo is a virtual package provided by:\n  php7.3-common 7.3.11-1~deb10u1\n  php7.2-common 7.2.9-1+b2\n  php7.1-common 7.1.20-1+b2\n\nPackage php-iconv is a virtual package provided by:\n  php7.3-common 7.3.11-1~deb10u1\n  php7.2-common 7.2.9-1+b2\n  php7.1-common 7.1.20-1+b2\n\nPackage php-posix is a virtual package provided by:\n  php7.3-common 7.3.11-1~deb10u1\n  php7.2-common 7.2.9-1+b2\n  php7.1-common 7.1.20-1+b2\n\nPackage php-dom is a virtual package provided by:\n  php7.3-xml 7.3.11-1~deb10u1\n  php7.2-xml 7.2.9-1+b2\n  php7.1-xml 7.1.20-1+b2\n\nPackage php-simplexml is a virtual package provided by:\n  php7.3-xml 7.3.11-1~deb10u1\n  php7.2-xml 7.2.9-1+b2\n  php7.1-xml 7.1.20-1+b2\n\nPackage php-xmlreader is a virtual package provided by:\n  php7.3-xml 7.3.11-1~deb10u1\n  php7.2-xml 7.2.9-1+b2\n  php7.1-xml 7.1.20-1+b2\n\nPackage php-xmlwriter is a virtual package provided by:\n  php7.3-xml 7.3.11-1~deb10u1\n  php7.2-xml 7.2.9-1+b2\n  php7.1-xml 7.1.20-1+b2\n\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "Package php-ctype is a virtual package provided by:", "  php7.3-common 7.3.11-1~deb10u1", "  php7.2-common 7.2.9-1+b2", "  php7.1-common 7.1.20-1+b2", "", "Package php-exif is a virtual package provided by:", "  php7.3-common 7.3.11-1~deb10u1", "  php7.2-common 7.2.9-1+b2", "  php7.1-common 7.1.20-1+b2", "", "Package php-fileinfo is a virtual package provided by:", "  php7.3-common 7.3.11-1~deb10u1", "  php7.2-common 7.2.9-1+b2", "  php7.1-common 7.1.20-1+b2", "", "Package php-iconv is a virtual package provided by:", "  php7.3-common 7.3.11-1~deb10u1", "  php7.2-common 7.2.9-1+b2", "  php7.1-common 7.1.20-1+b2", "", "Package php-posix is a virtual package provided by:", "  php7.3-common 7.3.11-1~deb10u1", "  php7.2-common 7.2.9-1+b2", "  php7.1-common 7.1.20-1+b2", "", "Package php-dom is a virtual package provided by:", "  php7.3-xml 7.3.11-1~deb10u1", "  php7.2-xml 7.2.9-1+b2", "  php7.1-xml 7.1.20-1+b2", "", "Package php-simplexml is a virtual package provided by:", "  php7.3-xml 7.3.11-1~deb10u1", "  php7.2-xml 7.2.9-1+b2", "  php7.1-xml 7.1.20-1+b2", "", "Package php-xmlreader is a virtual package provided by:", "  php7.3-xml 7.3.11-1~deb10u1", "  php7.2-xml 7.2.9-1+b2", "  php7.1-xml 7.1.20-1+b2", "", "Package php-xmlwriter is a virtual package provided by:", "  php7.3-xml 7.3.11-1~deb10u1", "  php7.2-xml 7.2.9-1+b2", "  php7.1-xml 7.1.20-1+b2", ""]}

**Comment by holta** [comment]
@m-anish please take a closer look (this is looking a whole heckuva lot better).  I'd like to merge this very soon, and then we can build on Nextcloud-on-NGINX from there.

Do you see any immediate fixes needed herein?

And going forward...which Nextcloud-on-NGINX improvements are most needed & most attainable to make our community materially open to newcomers?

e.g. in addition to some of the things you mentioned at the top of #2119, here's one ugly little bug we should try to nail soon if you (or @pdey!) have ideas?

- Nextcloud's out-of-the-box promo video fails to render, because URL's like the following...
  - http://box/nextcloud/apps/firstrunwizard/img/Nextcloud.webm
  - http://box/nextcloud/apps/firstrunwizard/img/Nextcloud.mp4
...are all redirected to... http://box/nextcloud/apps/files/
...which in turn redirects to... http://box/nextcloud/apps/files/?dir=/&fileid=2

(Here's the latest [nextcloud-nginx.conf](https://github.com/holta/iiab/blob/nextcloud-native-on-nginx/roles/nextcloud/templates/nextcloud-nginx.conf.j2) in question if you have ideas!)

**Comment by holta** [comment]
Smoke-tested & functionality-tested with a fresh install onto Ubuntu 18.04.

In addition to testing on the 3 other OS's mentioned above (Ubuntu 19.10, Debian 10.3, Raspbian Lite).

File: roles/nextcloud/tasks/install.yml [hunk 1]
```diff
@@ -1,183 +1,117 @@
...
+      - php-exif         # Optional (for image rotation in pictures app)
+      - php-fileinfo     # Optional (enhances file analysis performance)
+      - php-iconv
+      - php-posix
+      - php-simplexml
+      - php-xmlreader
+      - php-xmlwriter
     state: present
-  when: is_debuntu | bool
+  when: not is_raspbian
```
Comment (jvonau) [inline roles/nextcloud/tasks/install.yml:88]: Given you are seeing 
php-exif is a virtual package provided by:\n php7.3-common 7.3.11-1deb10u1\n
php-fileinfo is a virtual package provided by:\n php7.3-common 7.3.11deb10u1\n
php-ctype is a virtual package provided by:\n php7.3-common 7.3.11-1deb10u1\n
as those modules are now part of common?? Perhaps that could be   
when: not {{ php_version }}| int < 7.3
or php_version is defined in os.yml so you could use the vars that are available there. 
Strange raspbian is not seeing the virtual depends correctly, pity.

Comment (holta) [reply roles/nextcloud/tasks/install.yml:88]: `php_version` is definitely 7.3 on Raspbian Stretch on RPi, as I've reconfirmed in places like:
https://github.com/iiab/iiab/blob/master/vars/raspbian-10.yml#L29

Does it make sense to add in `php-common` to the list at https://github.com/iiab/iiab/blob/master/roles/nextcloud/tasks/install.yml#L43-L72 to be safe?

Also...is it likely that most all 9 php-* packages at https://github.com/iiab/iiab/blob/master/roles/nextcloud/tasks/install.yml#L74-L88 are gratuitous, having been installed anyway by larger superset php-* packages...such that many or all of these 9 lines can be yanked?

Comment (jvonau) [reply roles/nextcloud/tasks/install.yml:88]: > pi@raspberrypi:~ $ sudo apt install php-ctype
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package php-ctype is a virtual package provided by:
  php7.3-common 7.3.11-1~deb10u1
  php7.2-common 7.2.9-1+b2
  php7.1-common 7.1.20-1+b2
You should explicitly select one to install.

> pi@raspberrypi:~ $ sudo apt install php-dom
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Package php-dom is a virtual package provided by:
  php7.3-xml 7.3.11-1~deb10u1
  php7.2-xml 7.2.9-1+b2
  php7.1-xml 7.1.20-1+b2
You should explicitly select one to install.

Think you went the wrong way in trying to unify the package names by dropping the use of php_version in the package name of the apt install request.

Comment (jvonau) [reply roles/nextcloud/tasks/install.yml:88]: well maybe not, php(7.3)-common does get dragged in.
> pi@raspberrypi:~ $ sudo apt install php-mysql 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  php-common php7.3-common php7.3-mysql
The following NEW packages will be installed:
  php-common php-mysql php7.3-common php7.3-mysql
0 upgraded, 4 newly installed, 0 to remove and 0 not upgraded.
Need to get 635 kB of archives.
After this operation, 6,878 kB of additional disk space will be used.
Do you want to continue? [Y/n] n
Abort.

 I think that list you have is a stale apt listing from an older php version.

Comment (holta) [reply roles/nextcloud/tasks/install.yml:88]: fyi this helpful discussion is moving / has moved to #2258.

**Comment by holta** [comment]
> PR #2280 implements these 2 for now:
> 
> * www_base
> * www_options

Done.  If further renaming is truly necessary, that's OK too!

But for now both names conveniently correspond to their parent roles {3-base-server, 4-server-options} which can't hurt...unless/until they adopt new parents at some future date :-)

**Pull request merged**